src/TensorNetwork.jl

```diff
@@ -53,25 +54,44 @@ struct TensorNetwork <: AbstractTensorNetwork
             dict
         end
 
-        # Check for inconsistent index dimensions
-        for ind in keys(indexmap)
-            dims = map(tensor -> size(tensor, ind), indexmap[ind])
-            length(unique(dims)) == 1 || throw(DimensionMismatch("Index $(ind) has inconsistent dimension: $(dims)"))
+        if check_index_sizes === nothing || check_index_sizes == true
+            for ind in keys(indexmap) # Check for inconsistent index dimensions
+                dims = map(tensor -> size(tensor, ind), indexmap[ind])
+                length(unique(dims)) == 1 ||
+                    throw(DimensionMismatch("Index $(ind) has inconsistent dimension: $(dims)"))
+            end
+
+            return new(indexmap, tensormap, CachedField{Vector{Tensor}}(), Ref(true))
         end
 
-        return new(indexmap, tensormap, CachedField{Vector{Tensor}}())
+        return new(indexmap, tensormap, CachedField{Vector{Tensor}}(), Ref(false))
     end
 end
 
 TensorNetwork() = TensorNetwork(Tensor[])
 TensorNetwork(tn::TensorNetwork) = tn
 
+function set_unsafe_context!(tn::TensorNetwork, uc::Bool)
+    tn.check_index_sizes[] = uc
+    return tn
+end
+
 """
     copy(tn::TensorNetwork)
 
 Return a shallow copy of a [`TensorNetwork`](@ref).
 """
-Base.copy(tn::TensorNetwork) = TensorNetwork(tensors(tn))
+function Base.copy(tn::TensorNetwork)
+    new_tn = TensorNetwork(tensors(tn); check_index_sizes=tn.check_index_sizes[])
+
+    # Check if there's an active UnsafeContext with tn in it
+    uc = current_unsafe_context()
```

Comment:
This is what I want to talk to you about tomorrow. We just need to make a decision between 2 options.